bootstrap/src/basis/file.mli

```diff
@@ -0,0 +1,122 @@
+open Rudiments
+
+module Path = String
+
+module Mode : sig
+  type t =
+  | R_O   (** open existing file for read, failing if it does not exist *)
+  | W     (** create new or truncate existing file and open for write *)
+  | W_A   (** create new or append to existing file and open for write *)
+  | W_AO  (** append to existing file and open for write, failing if it does not
+              exist *)
+  | W_C   (** create new file and open for write, failing if file exists *)
+  | W_O   (** truncate existing file and open for write, failing if file does
+              not exist *)
+  | RW    (** create new or truncate existing file and open for read and write *)
+  | RW_A  (** create new append to existing file and open for read and write *)
+  | RW_AO (** append to existing file and open for read and write, failing if
+              file does not exist *)
+  | RW_C  (** create new file and open for read and write, failing if file
+              exists *)
+  | RW_O  (** truncate existing file and open for read and write, failing if
+              file does not exist *)
+end
+
+type t
+(** mutable Unix file descriptor *)
+
+type e
+
+val of_path: ?mode:Mode.t -> Path.t -> (t, e) Result.t
+(** [of_path mode path] opens file in [mode] at [path] and returns the resulting
+    file or an error if file could not be opened. *)
+
+val of_path_hlt: ?mode:Mode.t -> Path.t -> t
+(** [of_path mode path] opens file in [mode] at [path] and returns the resulting
+    file or halts if file could not be opened. *)
+
+val read: ?n:usize -> t -> (Bytes.t, e) Result.t
+(** [read n t] reads up to [n] bytes from [t] and returns the read bytes or an
+    error if bytes could not be read. *)
+
+val read_hlt: ?n:usize -> t -> Bytes.t
+(** [read_hlt n t] reads up to [n] bytes from [t] and returns the read bytes or
+    halts if bytes could not be read. *)
+
+val write: Bytes.t -> t -> e option
```

Comment:
Maybe we need to support partial writes.  Ordinarily that just causes a bunch of trouble for files (yay interruptible syscalls), but what about for broken pipes, where we did write some bytes before the pipe was closed?  I think this comes up for sockets too.

```ocaml
val write: Bytes.t -> t -> (usize, e) Result.t
```

Reply:
You're right. Maybe partial writes should return an error that contain the number of lines written, and `None` still indicates success? Regardless, the unix standard write operation returns the number of bytes written, so we should as well.